feat: add lambdaless (#1)
tlvince/tlvince.github.io: _drafts/lambdaless.md

---
title: Lambdaless
abstract: Serverless architectures without Lambda
---

Lets assume you need to expose a JSON file behind an API. Using a serverless approach with AWS, you might first reach for an architecture like the following:

[![API Gateway to Lambda to S3][apiglth]][apigl]

[apigl]: /assets/img/api-gateway-to-lambda-to-s3.png
[apiglth]: /assets/img/th/api-gateway-to-lambda-to-s3.png 'API Gateway to Lambda to S3'

> ... i.e. an API Gateway in front of a Lambda, which calls S3. Alternatively, did you know you could remove the Lambda and have API Gateway call S3 directly?

Reviewer: ```suggestion
... i.e. an API Gateway in front of a Lambda, which calls S3. But did you know you could remove the Lambda and have API Gateway call S3 directly?
```

[integration type]: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-api-integration-types.html
[mapping templates]: https://docs.aws.amazon.com/apigateway/latest/developerguide/models-mappings.html
[velocity]: https://velocity.apache.org/engine/devel/vtl-reference.html
[jsonpath]: https://goessner.net/articles/JsonPath/

## Walkthrough

Continuing with the S3 example above, create an API Gateway with a GET method and set up the integration request per the following:

- choose the AWS service type, region and Simple Storage Service (S3)
- select the GET HTTP method
- select the "use path override" action type
- enter the object's `<bucket>/<prefix>` in the path override field

[![API Gateway S3 Integration Request][apigs3ireqth]][apigs3ireq]

[apigs3ireq]: /assets/img/api-gateway-s3-integration-request.png
[apigs3ireqth]: /assets/img/th/api-gateway-s3-integration-request.png 'API Gateway S3 Integration Request'

> Create an IAM role that has a policy that has `s3:GetObject` permission on your `<bucket>/<prefix>` and Trust Relationship that allows API Gateway to assume it and that's it; switch to the test view, click "test" and you should see the contents of your JSON object in the response body:

Reviewer: ```suggestion
Create an IAM role that has a policy that has `s3:GetObject` permission on your `<bucket>/<prefix>` and a Trust Relationship that allows the API Gateway to assume it to be so. Now all you need to do is switch to the test view, click "test" and you should see the contents of your JSON object in the response body:
```

      "L": [
        #foreach($item in $inputRoot.items)
        {
          "S": "$item"
        }#if($foreach.hasNext),#end
        #end
      ]
    },
    "createdAt": {
      "N": "$context.requestTimeEpoch"
    }
  }
}
```

[context]: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-mapping-template-reference.html#context-variables-template-example
[data type descriptors]: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.LowLevelAPI.html#Programming.LowLevelAPI.DataTypeDescriptors

### Other ideas

- use the `HTTP_PROXY` integration to bypass region-locked websites
- pump events into an SQS queue
- raise AWS Support tickets using your existing customer service solution

## Advantages

A simple Lambda may seem innocuous at first, but each function comes with their own maintenance cost including:

- maintaining the application code
- maintaining dependencies
- any CI/CD tooling around delivering that code
- performing runtime upgrades
- security scanning
- configuring monitoring and alerts (e.g. CloudWatch)
- configuring instrumentation (e.g. X-Ray)

Removing a Lambda means fewer resources to maintain, test and pay for.

> 

Reviewer: ```suggestion
```

> Latency is also reduced. There are less hops in the chain and the issue of cold starts disappears.

Reviewer: ```suggestion
Latency is also reduced and there are less hops in the chain and the issues with cold starts disappears.
```

## Disadvantages

> There are a number drawbacks to consider. Probably most apparent is the fact that you can only integrate with a single service. This limits the approach to simple integrations and rules out complex logic e.g. joins.

Reviewer: ```suggestion
There are however a number drawbacks to consider with this Lambdaless method. Probably most apparent is the fact that you can only integrate with a single service at a time. This limits the approach to simple integrations and rules out complex logic e.g. joins.
```

Velocity, whilst offering some level of [control flow][vtl-directives] such as `if/else` and loops, as well as AWS's own extensions such as [util functions][], is somewhat of a niche language and introduces its own complexity over using your Lambda runtime language of choice (e.g. JavaScript, Python).

> This approach is also tightly coupled with API Gateway. The `AWS` integration type and request/response mapping template approach is unique to API Gateway and therefore is less portable than Lambda application logic (which is easier to abstract from the Lambda environment itself).

Reviewer: ```suggestion
This method is also tightly coupled with the API Gateway. The `AWS` integration type and request/response mapping template approach is unique to API Gateway and therefore is less portable than Lambda application logic (which is easier to abstract from the Lambda environment itself).
```
